WordCount: Add types

nb opened the pull request:
Part of #18838.

Add types to `wordcount` package.

**Testing Instructions**

0. Run `npm run build:package-types` –– shouldn't be any errors.
1. Try it in the UI – open the sample post and it should say something around 717 words. And start a new post and it should count words properly.

I am quite new to TypeScript and adding types to Gutenberg, so for the sake of learning, bring all of your nitpicking :-)

nb committed:
Define WordCountStrategy

It’s a constant set of strings.

nb committed:
TypeScript config for wordcount

nb committed:
Define a type for all the settings fields

nb committed:
Return unmodified text if regexp is missing

For consistency – all the rest of the strip* functions work like this.

Those aren’t exported outside of the wordcount package, so should be
safe to change the semantics.

nb committed:
Use WordCountSettings type in all strip* functions

nb committed:
No need to bind with current this

Since the functions don't use this, we can safely use null. Also helps
with type-checking :-)

We still need the bind, even if it's not a class, because of `settings`.
Not sure if syntax like `text => stripTags( settings, text )` will be
more readable, considered it, but decided to liimt the surface of my
changes.

nb committed:
Switch helper functions to returning a number

This hides their complexity (including complex return types) inside
themselves and makes makes the final `count()` function simpler.

As for the `count()` function, I considered adding a TypeScript ignore,
though the chance for adding a fourth type and not dealing with it here
is real. Also, the three `if` clauses aren't so hard to read.

nb committed:
Update autogenerated docs

nb committed:
Add descriptions for WordCountSettingsFields

Honestly, mostly to make the linter happy, not sure how useful they are.

nb commented on packages/wordcount/src/stripConnectors.js:
```diff
@@ -1,7 +1,7 @@
  * Replaces items matched in the regex with spaces.
  *
- * @param {Object} settings The main settings object containing regular expressions
+ * @param {import("./index").WordCountSettings} settings The main settings object containing regular expressions
```
Since those are used in only one place, opted in for using `import` directly, without a `typedef`.

aduth replied:
> Since those are used in only one place, opted in for using `import` directly, without a `typedef`.

Personally I feel it's fine to make these compromises where sensible.

nb replied:
I obviously agree via making the compromise :) Resolving.

aduth commented on packages/wordcount/src/defaultSettings.js:
```diff
@@ -1,3 +1,32 @@
+ */
+
+/**
+ * @typedef {Object} WordCountSettingsFields
```
What I've generally observed is that the "Object" type is interpreted by TypeScript in a way which isn't expected, and often results in types not being enforced (i.e. inferred as `any`).

>Don’t ever use the types Number, String, Boolean, Symbol, or Object These types refer to non-primitive boxed objects that are almost never used appropriately in JavaScript code.

See also: https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#general-types

Through the course of discussion in #20427, it was found that `Record` tends to work pretty well where we might otherwise reach to `Object`.

Documentation: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#generic-types

In this case, and as noted elsewhere in the documentation, you can omit `{Object}` altogether for `@typedef` that you intend to ascribe properties:

>Note that there is no {Object} between @typedef and the type name. As @propertys below tells us that it is a type for objects, it is recommend to not use {Object} when you want to define types for your objects.

https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#custom-types

```suggestion
 * @typedef WordCountSettingsFields
```

nb replied:
Good catch, @aduth. I think this was remnant of another iteration and now it’s super obvious I glossed over the “Custom Types” section of the guidelines :) Fixed in 51e005b.

aduth commented on packages/wordcount/src/defaultSettings.js:
```diff
@@ -1,3 +1,32 @@
+/**
+ * @typedef {Object} WordCountSettingsFields
+ * @property {RegExp} HTMLRegExp Regular expression that matches HTML tags
+ * @property {RegExp} HTMLcommentRegExp Regular expression that matches HTML comments
+ * @property {RegExp} spaceRegExp Regular expression that matches spaces in HTML
```
It should be that these "columns" are aligned:

```suggestion
 * @property {RegExp} HTMLcommentRegExp Regular expression that matches HTML comments
 * @property {RegExp} spaceRegExp       Regular expression that matches spaces in HTML
```

Reference:

- https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/#aligning-comments
- https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#custom-types

Of course, it should account for all other properties as well, where it becomes pretty difficult, at least when also trying to adhere to a particular line length (e.g. 80 characters), and when types can be complex when inlined. Usually I try to balance between: Create `@typedef` with shorter names as appropriate in lieu of inline types like below `Partial<{type: WordCountStrategy, shortcodes: string[]}`, e.g. `Partial<WPTranslationMapping>`, vs. relenting on a hard line length (allow 100 or 120 per line within the DocBlock).

nb replied:
Thanks for the reminder, @aduth. An attempt at alignment lives in dc253c424d. To be honest, it's not more readable to me, I will leave it for you to decide how to proceed with the alignment :) I don't have a strong opinion.

P.S. To be honest he time spent on this makes me just use `wp-prettier` and stop caring about “readability” all together.

aduth replied:
> Thanks for the reminder, @aduth. An attempt at alignment lives in [dc253c4](https://github.com/WordPress/gutenberg/commit/dc253c424de4815bf7d7f50cdfd5974081415842). To be honest, it's not more readable to me, I will leave it for you to decide how to proceed with the alignment :) I don't have a strong opinion.

At least as guidelines currently stand, I think the changes you committed are good and should stay.

The general point is interesting, and it's one I'd be curious to raise for discussion at some point (e.g. [weekly JavaScript chat](https://make.wordpress.org/meetings/)). I'm personally conflicted. I feel there is some benefit to general readability in having the aligned columns. The problems stem largely on the end of _getting them to look that way_, between a combination of problems:

- The effort involved in applying the spacing. 
- The difficulty in choice of line length (phrasing, types choice and externalization, etc)

To the first: I definitely acknowledge that it's a huge pain, at least in VSCode. I recall it being a complete non-issue when I'd used Sublime Text + [DocBlockr](https://github.com/spadgos/sublime-jsdocs), which had very smart auto-indentation associated with the <kbd>Tab</kbd> key. I've yet to find something equivalent for VSCode. I even have a work-in-progress VSCode plugin on my local machine because of this exact problem. To be honest, it feels like something we're just lacking good tooling for, but becomes solved once it exists.

>P.S. To be honest he time spent on this makes me just use wp-prettier and stop caring about “readability” all together.

Well, we do have Prettier, but it does nothing for us for JSDoc 😄

aduth commented on packages/wordcount/src/defaultSettings.js:
```diff
@@ -1,3 +1,32 @@
+ */
+
+/**
+ * @typedef {Object} WordCountSettingsFields
```
It's currently promoted that our custom types be prefixed with `WP`:

```suggestion
 * @typedef {Object} WPWordCountSettingsFields
```

>A `WP` prefix should be applied to all custom types.

https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#custom-types

Personally, I've been waffling on this a bit lately. The general thinking was to avoid conflicts or possible ambiguity with types inherited from "global" namespace, where here without a prefix, there's a possibility it leaves room for misinterpreting this as some default type, vs. custom to our packages.

nb replied:
Oh, thanks, missing the _first_ section of the guidelines is almost embarrassing :) Fixed in d098351. 

P.S. VSCode's rename (F2 for me) worked great. When importing had to rename both types, though. I wish that `/** @typedef {import('./lib').Type} */` worked and implied the type name without having to repeat it.

aduth commented on packages/wordcount/src/index.js:
```diff
@@ -17,18 +17,30 @@ import stripShortcodes from './stripShortcodes';
 	const settings = extend( defaultSettings, userSettings );
 
-	settings.shortcodes = settings.l10n.shortcodes || {};
+	settings.shortcodes = settings.l10n?.shortcodes ?? [];
```
I'm surprised there is nothing flagged here, since as far as I can tell, `settings.l10n` should be considered required, no? Based on type `Partial<{type: WordCountStrategy, shortcodes: string[]}>`. Unless "Partial" means "possibly undefined", which could be possible but would be new information to me.

nb replied:
According to the documentation (and it seems to work like this in practice), Partial is exactly that – [allows for any of the properties to be missing](https://www.typescriptlang.org/v2/docs/handbook/utility-types.html#partialt). I am curious, what did you expect that `Partial` gave us?

aduth replied:
Oh, I understand that, but based on your usage of optional chaining on `l10n`, it would seem to imply that `settings.l10n` could be `undefined`. I cannot see how that could be the case, based on the type. The `Partial` is relevant for properties within `l10n`, e.g. `settings.l10n.shortcodes` could be omitted.

nb replied:
@aduth `WPWordCountSettings` is also a `Partial`:

`@typedef {Partial<WPWordCountSettingsFields>} WPWordCountSettings `

aduth replied:
> @aduth `WPWordCountSettings` is also a `Partial`:
> 
> `@typedef {Partial<WPWordCountSettingsFields>} WPWordCountSettings `

Ah, I failed to follow that. It makes sense now 👍 I guess this is another example where it's best to let TypeScript manage this and avoid the sort of human error I'm apparently prone to 😄

aduth commented on packages/wordcount/src/index.js:
```diff
@@ -50,56 +62,62 @@ function loadSettings( type, userSettings ) {
  * @param {string} text     The text being processed
- * @param {string} regex    The regular expression pattern being matched
- * @param {Object} settings Settings object containing regular expressions for each strip function
+ * @param {RegExp|undefined} regex    The regular expression pattern being matched
```
Currently, it is documented to use square-bracket notation for undefined values.

```suggestion
 * @param {RegExp} [regex]    The regular expression pattern being matched
```

>If a parameter is optional, use the square-bracket notation.

https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#nullable-undefined-and-void-types

This was left inherited / unchanged from upstream guidelines (the examples): https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/#functions

I've personally observed a few cases where TypeScript isn't interpreting the possibly-undefined type in ways I expect, where occasionally the ClosureCompiler format can work more effectively:

```suggestion
 * @param {RegExp=} regex    The regular expression pattern being matched
```

nb replied:
Thanks, @aduth, and I apologize for missing this comment the other day.

From what I have experienced, there is a subtle difference between a possible `undefined` value and an optional argument. In this case, I don't think we’d like to allow _omitting_ the `regex` argument, though it would be acceptable that it's supplied and its value is `undefined`.

One side effect I have observed when I tried out either `[regex]` or `{RegExp=}` was that VSCode didn't include the `regex` when automatically filling in function arguments:

<img width="854" alt="Screenshot 2020-05-06 at 15 03 24 " src="https://user-images.githubusercontent.com/27954/81174651-e3f64000-8faa-11ea-870f-e34804e69d91.png">

This specific case is especially weird, because the third argument is **not** optional – that makes the semantic difference between `Type|undefined` and `Type=` even more pronounced.

Based on all the above, let's keep `RegExp|undefined`.

sirreal replied:
Allowing optional parameters to precede required parameters is a bug that was recently fixed in TypeScript (unreleased): https://github.com/microsoft/TypeScript/issues/37703

This affected the `url` types: https://github.com/WordPress/gutenberg/pull/21926

---

I haven't looked closely, but do we want to allow `undefined` as the regex?

aduth commented on packages/wordcount/src/index.js:
```diff
@@ -50,56 +62,62 @@ function loadSettings( type, userSettings ) {
 	)( text );
 	text = text + '\n';
-	return text.match( regex );
+	return text.match( regex )?.length ?? 0;
```
Is it possible that `String#match` would result in a valid match object (with `length`) where `length` would be non-numeric?

sirreal replied:
Yep! I came across this when typing `autop` (thanks to TypeScript types, I had no idea). In that case, I added a type assertion:

https://github.com/WordPress/gutenberg/blob/a67c9a9c69d8daee389b5d89ef460db7718b90e8/packages/autop/src/index.js#L63-L67

> Return value
> An `Array` whose contents depend on the presence or absence of the global (`g`) flag, or `null` if no matches are found.
> - If the `g` flag is used, all results matching the complete regular expression will be returned, but capturing groups will not.
> - if the `g` flag is not used, only the first complete match and its related capturing groups are returned. In this case, the returned item will have additional properties as described below.

[from MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/match#Return_value)

In a nutshell:

```js
'foo'.match(/f/)
// ["f", index: 0, input: "foo", groups: undefined]
'foo'.match(/f/g) // match behaves differently with "g" flag
// ["f"]
```

It's possible to inspect the flags. With a configurable regex like this, we might throw and error or at least show a warning if a regex with the `g` flag is provided:

```js
if ( regex.flags.includes( 'g' ) ) {
  throw new Error( '💥');
}

nb replied:
@aduth [docs say](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/match) it returns `Array` or `null`.

aduth replied:
I guess the question for me is: Is the following not enough if it's an `Array` or `null`?

```suggestion
	return text.match( regex )?.length;
```

nb replied:
@aduth in this case wouldn't the function return `undefined` and not `0`? We want it to return `number`.

aduth replied:
Oh! This is yet another case where I eagerly look to the end of the chain of an optional chaining expression. I was considering the `??` to be operating exclusively on `length`, not considering that it's also accounting for the `?.` short-circuit.

Funny enough, this is the exact sort of "gotcha" I just finished documenting in #22029 😆 

>This is an easy trap to fall into when eagerly assuming that a result of `trim()` will always return a string value, overlooking the fact the optional chaining may have caused evaluation to abort earlier with a value of `undefined`.

aduth commented on packages/wordcount/src/index.js:
```diff
@@ -111,18 +129,19 @@ function matchCharacters( text, regex, settings ) {
-
-		return results ? results.length : 0;
+	if ( settings.type === 'characters_including_spaces' ) {
+		matchRegExp = settings.characters_including_spacesRegExp;
```
Are this and `characters_excluding_spacesRegExp` below new settings introduced as of this pull request? I observe that they don't adhere to naming guidelines:

https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/#naming-conventions

I might expect something like:

```suggestion
		matchRegExp = settings.charactersIncludingSpacesRegExp;
```

Or perhaps more legible:

```suggestion
		matchRegExp = settings.charactersIncludingSpacesPattern;
```

**Edit:** Re-reading the changes, I think I understand this now as not being introduced here. I was trying to search for the literal phrase "characters_including_spacesRegExp", but I see it was previously concatenated above as `settings[ type + 'RegExp' ]`. The advice above still holds true, but given this isn't a change introduced here and for backwards-compatibility commitment, I don't think we want to do anything.

nb replied:
Yeah… I almost refactored this portion, though `settings` de-facto being public interface stopped me. Maybe adding a comment makes sense, but I hope nobody will try to override this in the first place. Don't really want to make their lives easier and I would hope they re-consider :)

aduth commented on packages/wordcount/src/index.js:
```diff
@@ -111,18 +129,19 @@ function matchCharacters( text, regex, settings ) {
-		return 0;
+	const settings = loadSettings( type, userSettings );
+	let matchRegExp;
+	if ( settings.type === 'words' ) {
```
Depending on your feelings about `switch` statements, the function logic looks like a good candidate:

```js
switch ( settings.type ) {
	case 'words':
		// ...

	case 'characters_including_spaces':
		// ...

	// ...
}
```

nb replied:
I tried the `switch` and the `if`s looked more aesthetically pleasing to me in this specific case. However, it just occurred to me that we can skip the `break`s, because we will have a `return` in each case and that actually tips my aesthetic scales towards `switch`. Thank you! Done in 0ac3aff.

aduth commented on packages/wordcount/src/stripConnectors.js:
```diff
@@ -1,7 +1,7 @@
  * Replaces items matched in the regex with spaces.
  *
- * @param {Object} settings The main settings object containing regular expressions
+ * @param {import("./index").WordCountSettings} settings The main settings object containing regular expressions
```
I wish it could be automatically enforced via one of either ESLint (e.g. the JSDoc plugin in use) or by Prettier, but I'd think we should standardize quotes usage to use the same single-quote convention otherwise standard:

https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/#strings

nb replied:
Oh, that's my bad. Fixed in 244b8a131f. And agreed that it would’ve been nice for the tools to have taken care or warned me earlier.

aduth commented on packages/wordcount/src/stripSpaces.js:
```diff
@@ -10,4 +10,5 @@ export default function( settings, text ) {
 	if ( settings.spaceRegExp ) {
 		return text.replace( settings.spaceRegExp, ' ' );
 	}
+	return text;
```
It seems sensible to me. It has me somewhat concerned if there were no unit tests to verify this fallback!

Would be a nice improvement, though fine to consider out of scope.

nb replied:
Thanks, @aduth. My first thought when trying to cover this code with tests would be to actually redesign it, which I really wanted to avoid with this PR :)

For example, the possibility for `spaceRegExp` to be undefined or the option to be changed from the outside seem very much in [YAGNI land](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) for me and would consider eliminating it. But it works quite well right now, the public interface is reasonably tested, refactoring/rewriting is on the other side of the cost-benefit line for me.

nb commented:
Thank you for the thorough review and comments, @aduth – I really appreciate it 🤩 Will be looking through it and addressing the feedback in the next day or two.

nb committed:
Switch to single quotes

I totally forgot about the convention in this context. Of course, I
would have expected some of the tools to have mentioned it along the
way, but alas…

nb committed:
Inline import @typedef to single line

In order to do that the type descriptions moved to the source, instead
of some of them living next to the import.

One trade-off for me was that VSCode now doesn’t pick up the
descriptions of the imported types.

nb commented on packages/wordcount/src/defaultSettings.js:
```diff
@@ -1,3 +1,32 @@
+/**
+ * @typedef {import("./index").WordCountStrategy} WordCountStrategy
+ */
```
Thanks, @aduth, that's useful and tried it in 3126c5db43. One drawback for me is that VSCode doesn't pick up the descriptions of the imported types, but that may be something in my setup that I haven't figured out.

aduth replied:
> One drawback for me is that VSCode doesn't pick up the descriptions of the imported types

I could see that, since while in practice we use this as a sort of import / aliasing, in reality it's creating a brand new type which happens to inherit _its type_ from where it's imported. That may very well not include details like the description. I've also seen issues when using this pattern to "import" a [generic type](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#template), where the details of the generic are lost.

Unfortunately, I don't know there's any other option here, aside from inlining every `import` where it's used.

nb committed:
Switch to switch

Has better rhythm than the `if`s (for me, it’s subjective). I had tried
the `switch` before, but I didn’t think of skipping the `break`s. We
don’t need them with all the `returns.

Technically we don’t need the `return 0` since we’re exhausting all
options for `type` union. However, I’d rather err on the side of
caution for cases when the codebase isn’t type checked.

Props @aduth.

nb committed:
Prefix custom types with WP

See https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#custom-types

nb committed:
Remove unneeded {Object}

Props @aduth. See https://github.com/WordPress/gutenberg/blob/b73bb6def033d7bb6fe850a186e27ab5b12ce235/docs/contributors/coding-guidelines.md#custom-types

nb committed:
Align comments

See https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/#aligning-comments

For the record, because of some really long strings it’s not more 
readable to me.

nb commented:
@aduth the feedback should be address, feel free to have another look when you get a chance 🙇

sirreal commented on packages/wordcount/src/defaultSettings.js:
```diff
@@ -1,3 +1,34 @@
+ * @typedef {Partial<WPWordCountSettingsFields>} WPWordCountSettings
+ */
+
+/**
+ * @type {WPWordCountSettings}
+ */
```
The inferred type may be better than the explicit type here. I'd consider omitting this:

```suggestion
```

With the annotation, all the properties are the `Partial<…>` type, so everything is possibly `undefined`. If we let the type be inferred, it can determine the type pretty well.

---

In TypeScript, I'd also consider a [`const` assertion](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions), but that seems to be unavailable via JSDoc (https://github.com/microsoft/TypeScript/issues/30445). We might be able to get something similar by using `Object.freeze` (for a runtime cost) or by playing with `@readonly` JSDoc tags on properties (not sure how well this works or applies).

nb replied:
Thanks, @sirreal – done in ffb27cc. It did help with the narrowing the type.

sirreal commented on packages/wordcount/src/index.js:
```diff
@@ -17,18 +17,26 @@ import stripShortcodes from './stripShortcodes';
+ * @param {WPWordCountSettings} userSettings Custom settings for the count.
  *
- * @return {void|Object|*} The combined settings object to be used.
+ * @return {WPWordCountSettings} The combined settings object to be used.
```
It seems appealing to have a narrower return type here. The `Partial<…>` type is good for the inputs, all of the fields are optional. However, we have better guarantees on the return type. Something like `{WPWordCountSettings & typeof defaultSettings}` might be a good start, that _should_ be the type of `const settings = extend( defaultSettings, userSettings )` a few lines below.

nb replied:
Thanks, @sirreal – see ffb27cc and 4576c98 for an exploration in this area.

sirreal reviewed (commented):
I've left a few thoughts for consideration, nothing major or blocking. Thanks for joining us on this typed adventure, @nb! Great to have you along 😁

nb committed:
Introduce a new, narrower, settings type

The idea came from @sirreal ([0], [1]) to use the already defined keys
in defaultSettings as a natural way to have a stronger type guarantee.

Once we have the narrower type, we can distinguish between settings
object we expect from the user, where everything is optional and a
settings object that comes from `loadSettings()` where we know that a
massive amount of the properties have been set as part of `defaultSettings`.

[0] https://git.io/JfVWj
[1] https://git.io/JfVle

nb committed:
Based on previous commit, relax undefined checks

Since we have a typecheck guarantee that most properties are there,
we don’t need to do all those checks.

Note that we still do the checks for `shortcodesRegExp` and `type`. Since
the guarantee for them doesn’t come from `defaultSettings`, but from
`loadSettings()`, we would have needed yet another layer of type to
describe the settings coming from `defaultSettings` + `type`, `shortcodes`,
and `shortcodesRegExp`.

Personally, this crossed a line, and I decided
that leaving the checks in `stripShortcodes.js` and the `default` in the
`switch` in `count()` were a smaller price to pay than adding yet another
type and dealing with type trickery (would’ve needed a `Pick` and to
disable `jsdoc/valid-types` again).

nb commented on packages/wordcount/src/index.js:
```diff
@@ -111,18 +125,19 @@ function matchCharacters( text, regex, settings ) {
  */
 
```
Thanks, @sirreal, fixed in ffb27cc. Sneaked it in :)

sirreal commented on packages/wordcount/src/index.js:
```diff
@@ -17,18 +17,29 @@ import stripShortcodes from './stripShortcodes';
+ * @return {WPWordCountSettings} The combined settings object to be used.
  */
 function loadSettings( type, userSettings ) {
 	const settings = extend( defaultSettings, userSettings );
```
This looks like a (pre-existing) bug, `extend` is an alias for [`assignIn`](https://lodash.com/docs/4.17.15#assignIn):

> This method is like _.assign except that it iterates over own and inherited source properties.
> 
> Note: This method mutates object.

We shouldn't be mutating shared default settings 😱 

```suggestion
	const settings = extend( {}, defaultSettings, userSettings );
```

That's grounds for it's own PR, but since we're deep into a significant refactor, we might sneak it in 🙂

sirreal commented on packages/wordcount/src/index.js:
```diff
@@ -37,7 +48,7 @@ function loadSettings( type, userSettings ) {
 	}
 
-	settings.type = type || settings.l10n.type;
+	settings.type = type || settings.l10n?.type;
```
`type` is not an optional argument, it should be one of the valid strings. I suspect we can simplify things here.

nb replied:
Oh, yeah, thanks.Fixed in aa335aa.

sirreal commented on packages/wordcount/src/defaultSettings.js:
```diff
@@ -1,3 +1,40 @@
+ * @typedef {Partial<WPWordCountSettingsFields>} WPWordCountUserSettings
+ */
+
+/**
+ * Word counting settings that include non-optional values we set if missing
+ *
+ * @typedef {WPWordCountUserSettings & typeof defaultSettings} WPWordCountDefaultSettings
+ */
+/* eslint-enable jsdoc/valid-types */
```
I would like to avoid this type completely. I believe we can get to a place where we work with a fairly strict interface that's already defined as `WPWordCountSettingsFields`. `loadSettings` should be able to return a well typed interface on its inference of `defaultSettings` and `userSettings`.

nb replied:
Hm, I agree on principle that it would be nice if we had one less type, @sirreal. At the same time, if we want to commit to always returning a full interface we will need to commit to firmer decisions around `l10n` and `shortcodes` – their goal is bluerrier to me at this point and the risk outweighs my desire for purity and simplicity at this point. If you’d like to take a stab at it I’d support you with a review :)

sirreal replied:
This was part of what I did in the diff I shared [here](https://github.com/WordPress/gutenberg/pull/22077#issuecomment-634164997), but I'm happy to drop that for now.

sirreal reviewed (commented):
I really like where you've taken this! It's so nice to see where we can refactor the internal interfaces supported by the type system. I've left some notes and comments, I'll follow up with a larger comment and a diff of some changes I've been exploring.

Thanks for continuing to iterate here!

sirreal commented:
I'll share some work I did exploring some ideas I alluded to in my review https://github.com/WordPress/gutenberg/pull/22077#pullrequestreview-417759463.

Some notes:
- Use a strict settings interface, ignore the shape of `defaultSettings` and let type inference determine its shape.
- Simplify `loadSettings`. `loadSettings` is an internal function and we can make pretty strong assumptions about the values it receives.
- This change includes some defensive type/value juggling for `type` in the public `count` function. I explored this approach because, while our _types_ say that `type` is one of 3 string values, this is a public JavaScript API, and I don't know how much we can trust it 🙂.
  This approach allows us to simplify the internal `loadSettings` method and have nice type guarantees.

```js
/**
 * @param {string}                  text         The text being processed
 * @param {WPWordCountStrategy}     type	     The type of count. Accepts 'words', 'characters_excluding_spaces', or 'characters_including_spaces'.
 * @param {WPWordCountUserSettings} userSettings Custom settings object.
 * @return {number} The word or character count.
 */
export function count( text, type, userSettings ) {

	// Here, the `type` parameter has a strict type:
	// 'words' | 'characters_excluding_spaces' | 'characters_including_spaces'
	// But this is the JavaScript public API… we don't _really_ know what we get.

	// We _do_ want our public interface to claim the correct type, however!
	// Widening the public interface to `string` is undesirable.

	// We can widen this type internally:
	/** @type {string} */
	const userType = type;

	// Then we let the type system go from the wider `string` type back to our well known type.
	const validType =
		userType === 'characters_excluding_spaces' ||
		userType === 'characters_including_spaces'
			? userType
			: 'words';

	// Now our strict `loadSettings` function happily accepts our well-typed `validType`.
	const settings = loadSettings( validType, userSettings );
	// snip…
}
```

This is exploratory and I'm happy to discuss further. There are already some very nice changes here and I'm happy to do a more narrow review if you'd prefer. You did ask for all the nitpicking 🙂 

<details>
<summary>Diff without space</summary>

```diff
diff --git a/packages/wordcount/src/defaultSettings.js b/packages/wordcount/src/defaultSettings.js
index 020f44162..90394832c 100644
--- a/packages/wordcount/src/defaultSettings.js
+++ b/packages/wordcount/src/defaultSettings.js
@@ -1,9 +1,13 @@
 /** @typedef {import('./index').WPWordCountStrategy} WPWordCountStrategy */
 
-/** @typedef {Partial<{type: WPWordCountStrategy, shortcodes: string[]}>} WPWordCountL10n */
+/**
+ * @typedef WPWordCountL10n
+ * @property {string[]} shortcodes Localized shortcodes.
+ */
 
 /**
- * @typedef WPWordCountSettingsFields
+ * @typedef WPWordCountSettings
+ *
  * @property {RegExp}                    HTMLRegExp                        Regular expression that matches HTML tags
  * @property {RegExp}                    HTMLcommentRegExp                 Regular expression that matches HTML comments
  * @property {RegExp}                    spaceRegExp                       Regular expression that matches spaces in HTML
@@ -14,27 +18,18 @@
  * @property {RegExp}                    wordsRegExp                       Regular expression that matches words
  * @property {RegExp}                    characters_excluding_spacesRegExp Regular expression that matches characters excluding spaces
  * @property {RegExp}                    characters_including_spacesRegExp Regular expression that matches characters including spaces
- * @property {RegExp}              shortcodesRegExp                  Regular expression that matches WordPress shortcodes
+ * @property {RegExp|undefined}          [shortcodesRegExp]                Regular expression that matches WordPress shortcodes
  * @property {string[]}                  shortcodes                        List of all shortcodes
  * @property {WPWordCountStrategy}       type                              Describes what and how are we counting
- * @property {WPWordCountL10n}     l10n                              Object with human translations
+ * @property {WPWordCountL10n|undefined} [l10n]                            Object with human translations
  */
 
-// Disable reason: JSDoc linter doesn't seem to parse the union (`&`) correctly: https://github.com/jsdoc/jsdoc/issues/1285
-/* eslint-disable jsdoc/valid-types */
 /**
  * Lower-level settings for word counting that can be overridden.
  *
- * @typedef {Partial<WPWordCountSettingsFields>} WPWordCountUserSettings
+ * @typedef {Partial<WPWordCountSettings>} WPWordCountUserSettings
  */
 
-/**
- * Word counting settings that include non-optional values we set if missing
- *
- * @typedef {WPWordCountUserSettings & typeof defaultSettings} WPWordCountDefaultSettings
- */
-/* eslint-enable jsdoc/valid-types */
-
 export const defaultSettings = {
 	HTMLRegExp: /<\/?[a-z][^>]*?>/gi,
 	HTMLcommentRegExp: /<!--[\s\S]*?-->/g,
@@ -108,7 +103,4 @@ export const defaultSettings = {
 	 * \u2029 = paragraph separator
 	 */
 	characters_including_spacesRegExp: /[^\f\n\r\t\v\u00AD\u2028\u2029]/g,
-	l10n: {
-		type: 'words',
-	},
 };
diff --git a/packages/wordcount/src/index.js b/packages/wordcount/src/index.js
index f1ccce2f7..eceb8d3c9 100644
--- a/packages/wordcount/src/index.js
+++ b/packages/wordcount/src/index.js
@@ -18,7 +18,7 @@ import stripSpaces from './stripSpaces';
 import transposeHTMLEntitiesToCountableChars from './transposeHTMLEntitiesToCountableChars';
 
 /**
- * @typedef {import('./defaultSettings').WPWordCountDefaultSettings}  WPWordCountSettings
+ * @typedef {import('./defaultSettings').WPWordCountSettings}     WPWordCountSettings
  * @typedef {import('./defaultSettings').WPWordCountUserSettings} WPWordCountUserSettings
  */
 
@@ -37,26 +37,18 @@ import transposeHTMLEntitiesToCountableChars from './transposeHTMLEntitiesToCoun
  * @return {WPWordCountSettings} The combined settings object to be used.
  */
 function loadSettings( type, userSettings ) {
-	const settings = extend( defaultSettings, userSettings );
+	const settings = extend( {}, defaultSettings, userSettings, {
+		shortcodes: userSettings.l10n?.shortcodes ?? [],
+		type,
+	} );
 
-	settings.shortcodes = settings.l10n?.shortcodes ?? [];
-
-	if ( settings.shortcodes && settings.shortcodes.length ) {
+	if ( settings.shortcodes?.length ) {
 		settings.shortcodesRegExp = new RegExp(
 			'\\[\\/?(?:' + settings.shortcodes.join( '|' ) + ')[^\\]]*?\\]',
 			'g'
 		);
 	}
 
-	settings.type = type || settings.l10n?.type;
-
-	if (
-		settings.type !== 'characters_excluding_spaces' &&
-		settings.type !== 'characters_including_spaces'
-	) {
-		settings.type = 'words';
-	}
-
 	return settings;
 }
 
@@ -121,7 +113,16 @@ function countCharacters( text, regex, settings ) {
  * @return {number} The word or character count.
  */
 export function count( text, type, userSettings ) {
-	const settings = loadSettings( type, userSettings );
+	/** @type {string} */
+	const userType = type;
+
+	const validType =
+		userType === 'characters_excluding_spaces' ||
+		userType === 'characters_including_spaces'
+			? userType
+			: 'words';
+
+	const settings = loadSettings( validType, userSettings );
 	let matchRegExp;
 	switch ( settings.type ) {
 		case 'words':
```

</details>

<details>
<summary>Full diff</summary>

```diff
diff --git a/packages/wordcount/src/defaultSettings.js b/packages/wordcount/src/defaultSettings.js
index 020f44162..90394832c 100644
--- a/packages/wordcount/src/defaultSettings.js
+++ b/packages/wordcount/src/defaultSettings.js
@@ -1,40 +1,35 @@
 /** @typedef {import('./index').WPWordCountStrategy} WPWordCountStrategy */
 
-/** @typedef {Partial<{type: WPWordCountStrategy, shortcodes: string[]}>} WPWordCountL10n */
-
 /**
- * @typedef WPWordCountSettingsFields
- * @property {RegExp}              HTMLRegExp                        Regular expression that matches HTML tags
- * @property {RegExp}              HTMLcommentRegExp                 Regular expression that matches HTML comments
- * @property {RegExp}              spaceRegExp                       Regular expression that matches spaces in HTML
- * @property {RegExp}              HTMLEntityRegExp                  Regular expression that matches HTML entities
- * @property {RegExp}              connectorRegExp                   Regular expression that matches word connectors, like em-dash
- * @property {RegExp}              removeRegExp                      Regular expression that matches various characters to be removed when counting
- * @property {RegExp}              astralRegExp                      Regular expression that matches astral UTF-16 code points
- * @property {RegExp}              wordsRegExp                       Regular expression that matches words
- * @property {RegExp}              characters_excluding_spacesRegExp Regular expression that matches characters excluding spaces
- * @property {RegExp}              characters_including_spacesRegExp Regular expression that matches characters including spaces
- * @property {RegExp}              shortcodesRegExp                  Regular expression that matches WordPress shortcodes
- * @property {string[]}            shortcodes                        List of all shortcodes
- * @property {WPWordCountStrategy} type                              Describes what and how are we counting
- * @property {WPWordCountL10n}     l10n                              Object with human translations
+ * @typedef WPWordCountL10n
+ * @property {string[]} shortcodes Localized shortcodes.
+ */
+
+/**
+ * @typedef WPWordCountSettings
+ *
+ * @property {RegExp}                    HTMLRegExp                        Regular expression that matches HTML tags
+ * @property {RegExp}                    HTMLcommentRegExp                 Regular expression that matches HTML comments
+ * @property {RegExp}                    spaceRegExp                       Regular expression that matches spaces in HTML
+ * @property {RegExp}                    HTMLEntityRegExp                  Regular expression that matches HTML entities
+ * @property {RegExp}                    connectorRegExp                   Regular expression that matches word connectors, like em-dash
+ * @property {RegExp}                    removeRegExp                      Regular expression that matches various characters to be removed when counting
+ * @property {RegExp}                    astralRegExp                      Regular expression that matches astral UTF-16 code points
+ * @property {RegExp}                    wordsRegExp                       Regular expression that matches words
+ * @property {RegExp}                    characters_excluding_spacesRegExp Regular expression that matches characters excluding spaces
+ * @property {RegExp}                    characters_including_spacesRegExp Regular expression that matches characters including spaces
+ * @property {RegExp|undefined}          [shortcodesRegExp]                Regular expression that matches WordPress shortcodes
+ * @property {string[]}                  shortcodes                        List of all shortcodes
+ * @property {WPWordCountStrategy}       type                              Describes what and how are we counting
+ * @property {WPWordCountL10n|undefined} [l10n]                            Object with human translations
  */
 
-// Disable reason: JSDoc linter doesn't seem to parse the union (`&`) correctly: https://github.com/jsdoc/jsdoc/issues/1285
-/* eslint-disable jsdoc/valid-types */
 /**
  * Lower-level settings for word counting that can be overridden.
  *
- * @typedef {Partial<WPWordCountSettingsFields>} WPWordCountUserSettings
+ * @typedef {Partial<WPWordCountSettings>} WPWordCountUserSettings
  */
 
-/**
- * Word counting settings that include non-optional values we set if missing
- *
- * @typedef {WPWordCountUserSettings & typeof defaultSettings} WPWordCountDefaultSettings
- */
-/* eslint-enable jsdoc/valid-types */
-
 export const defaultSettings = {
 	HTMLRegExp: /<\/?[a-z][^>]*?>/gi,
 	HTMLcommentRegExp: /<!--[\s\S]*?-->/g,
@@ -108,7 +103,4 @@ export const defaultSettings = {
 	 * \u2029 = paragraph separator
 	 */
 	characters_including_spacesRegExp: /[^\f\n\r\t\v\u00AD\u2028\u2029]/g,
-	l10n: {
-		type: 'words',
-	},
 };
diff --git a/packages/wordcount/src/index.js b/packages/wordcount/src/index.js
index f1ccce2f7..eceb8d3c9 100644
--- a/packages/wordcount/src/index.js
+++ b/packages/wordcount/src/index.js
@@ -18,8 +18,8 @@ import stripSpaces from './stripSpaces';
 import transposeHTMLEntitiesToCountableChars from './transposeHTMLEntitiesToCountableChars';
 
 /**
- * @typedef {import('./defaultSettings').WPWordCountDefaultSettings}  WPWordCountSettings
- * @typedef {import('./defaultSettings').WPWordCountUserSettings}     WPWordCountUserSettings
+ * @typedef {import('./defaultSettings').WPWordCountSettings}     WPWordCountSettings
+ * @typedef {import('./defaultSettings').WPWordCountUserSettings} WPWordCountUserSettings
  */
 
 /**
@@ -37,26 +37,18 @@ import transposeHTMLEntitiesToCountableChars from './transposeHTMLEntitiesToCoun
  * @return {WPWordCountSettings} The combined settings object to be used.
  */
 function loadSettings( type, userSettings ) {
-	const settings = extend( defaultSettings, userSettings );
+	const settings = extend( {}, defaultSettings, userSettings, {
+		shortcodes: userSettings.l10n?.shortcodes ?? [],
+		type,
+	} );
 
-	settings.shortcodes = settings.l10n?.shortcodes ?? [];
-
-	if ( settings.shortcodes && settings.shortcodes.length ) {
+	if ( settings.shortcodes?.length ) {
 		settings.shortcodesRegExp = new RegExp(
 			'\\[\\/?(?:' + settings.shortcodes.join( '|' ) + ')[^\\]]*?\\]',
 			'g'
 		);
 	}
 
-	settings.type = type || settings.l10n?.type;
-
-	if (
-		settings.type !== 'characters_excluding_spaces' &&
-		settings.type !== 'characters_including_spaces'
-	) {
-		settings.type = 'words';
-	}
-
 	return settings;
 }
 
@@ -121,7 +113,16 @@ function countCharacters( text, regex, settings ) {
  * @return {number} The word or character count.
  */
 export function count( text, type, userSettings ) {
-	const settings = loadSettings( type, userSettings );
+	/** @type {string} */
+	const userType = type;
+
+	const validType =
+		userType === 'characters_excluding_spaces' ||
+		userType === 'characters_including_spaces'
+			? userType
+			: 'words';
+
+	const settings = loadSettings( validType, userSettings );
 	let matchRegExp;
 	switch ( settings.type ) {
 		case 'words':
```

</details>

nb committed:
Move eslint-disable rule to correct place

Doesn’t change functionality much, but it was more confusing spanning
multiple declarations.

nb commented on packages/wordcount/src/defaultSettings.js:
```diff
@@ -1,3 +1,40 @@
+ */
+
+// Disable reason: JSDoc linter doesn't seem to parse the union (`&`) correctly: https://github.com/jsdoc/jsdoc/issues/1285
+/* eslint-disable jsdoc/valid-types */
```
Thanks, addressed in 015caa5.

nb commented on packages/wordcount/src/defaultSettings.js:
```diff
@@ -1,3 +1,40 @@
+/**
+ * @typedef WPWordCountSettingsFields
+ * @property {RegExp}              HTMLRegExp                        Regular expression that matches HTML tags
+ * @property {RegExp}              HTMLcommentRegExp                 Regular expression that matches HTML comments
```
Since this is almost public interface, I am hesitant to change it here.

nb committed:
Don't mutate default settings object when extending it

`extend` is an alias for [assignIn](https://lodash.com/docs/4.17.15#assignIn):

> This method is like _.assign except that it iterates over own and inherited source properties.
> Note: This method mutates object.

Props @sirreal.

Co-authored-by: Jon Surrell <jon.surrell@automattic.com>

nb committed:
No need for fallback for required and validate arg

We know that `type` can’t be missing or falsy, so we can skip the check.

nb commented:
@sirreal I am having a hard time understanding what problem does this diff solve? I see that we simplified one function for the sake of moving the same logic to another at the same level of abstraction (even if one level down in indirection). We had the extra safety for the `type` argument before, it just lived in `loadSettings` and latter we used `settings.type`.

sirreal commented:
Thanks for your patience, I've had a lot come up that has prevented me from giving this the focus it deserves.

I'll try to clarify my reasoning, but I'll preface a lengthy response by saying that, after additional scrutiny, _it doesn't work as well as I expected_ 🙂 

I'll elaborate…

> I am having a hard time understanding what problem does this diff solve?

My goals were:

- Describe a single interface for settings.
- The interface is described by our type, not by the implementation of defaults (no `typeof defaultSettings`).
- We leverage the single interface to check:
  - The settings passed around internally in the package.
  - The shape of `defaultSettings` based on its inferred type.

The significant improvement I see is that we describe a single interface and use it to control several types. Rather than `defaultSettings` widening the Settings type used throughout the package (`Settings & typeof defaultSettings`), we use `Settings` and `loadSettings` the ensure that the inferred `defaultSettings` type is what we expect (it fits `Settings` and guarantees that we provide the required values).

Unfortunately, while this appeared to work very well in the [implementation I shared](https://github.com/WordPress/gutenberg/pull/22077#issuecomment-634164997), [I've continued testing and the type system doesn't provide the safety I'd hoped](https://codesandbox.io/s/agitated-lederberg-q12eh?file=/src/index.tsx). For example:

```ts
import { extend } from "lodash";

export type Options = {
  type: string;
  enabled: boolean;
};

// ⚠️ This type includes `{ enabled: string }`, it should be incompatible with `Options`
const defaultOptions = {
  type: "default-type",
  enabled: "baz"
};

// I'd expect a type error here, but there are none 😫
export function parseOptions(options: Partial<Options>): Options {
  return extend({}, defaultOptions, options);
}
```

The crux of this seems to be that `extend` (or `Object.assign` and similar) returns an intersection of the provided objects:

```ts
function assignIn<TObject, TSource1, TSource2>(
  object: TObject,
  source1: TSource1,
  source2: TSource2
): TObject & TSource1 & TSource2;
```

It's surprising to me that (irrelevant properties omitted) `{ enabled: string } & { enabled?: boolean }` somehow satisfies the return type `{ enabled: boolean }`. In my testing `{ enabled: string } & { enabled: boolean }` results in `{ enabled: never }`.

This may be related to https://github.com/microsoft/TypeScript/issues/12936.

sirreal commented:
> I see that we simplified one function for the sake of moving the same logic to another at the same level of abstraction (even if one level down in indirection).

Regarding `type`, my ideas are about when our internal type safety meets the public API.

> We had the extra safety for the `type` argument before, it just lived in `loadSettings` and latter we used `settings.type`.

We did:

https://github.com/WordPress/gutenberg/blob/aa335aa450b9faad587e07b978097a3da1f5086f/packages/wordcount/src/index.js#L51-L58

However, the type of `type` in this function is already `'words' | 'characters_excluding_spaces' | 'characters_including_spaces'`, which makes the block look redundant. The same is true of `type` in `count`, but I proposed explicitly widening and narrowing it. The same technique could also be applied in `loadSettings`.

<details>
<summary>Code sketching</summary>

This will log:

```
If you see this, something unexpected happened… 
Unsafe: foo 
Safe: B
```

TypeScript module:
```ts
type CountType = "A" | "B";

interface Settings {
  type: CountType;
}

// Public API
// This is great as long as it's invoked by well-typed code.
// However, we have very weak guarantees about JavaScript consumers.
export function takesType(type: CountType): CountType {
  // Our type is already infected if type isn't actually CountType 😱
  const s: Settings = makeSettings(type);

  switch (s.type) {
    case "A":
    case "B":
      return typeFromSettings(s);
  }

  // We're don't need to provide a return here according to the type system because all possible branches have returns!
  // This section is flagged as unreachable!
  // I'll include a bit more here…
  console.log("If you see this, something unexpected happened…");
  return typeFromSettings(s);
}

// Public API still claims type is CountType
export function safeTakesType(type: CountType): CountType {
  // Here, we're defensive. We don't trust the user provided `type`, so we cast to `unknown`
  // and rely on the type system to narrow the type.
  const userType: unknown = type;
  const safeType = userType === "A" ? userType : "B";

  const s: Settings = makeSettings(safeType);

  switch (s.type) {
    case "A":
    case "B":
      return typeFromSettings(s);
  }
}

function makeSettings(type: CountType): Settings {
  return { type };
}

function typeFromSettings(s: Settings): CountType {
  return s.type;
}
```

Consumer (JavaScript)
```js
import { safeTakesType, takesType } from "./type";

console.log("Unsafe:", takesType("foo"));
console.log("Safe:", safeTakesType("foo"));
```

</details>

sirreal reviewed (approved):
Thanks @nb! The changes are nice and I'm looking forward to ticking another package off the "typed" list.

This does need a rebase, but the changes look good and the tests continue to pass without modifications. All the code changes seem to be for the better 🙂 

I engaged in more academic exercises in this review than I usually would and I appreciate your patience, I learned a few things and I hope it was informative for you as well. None of my recent comments are blocking; they're part of the investigative and exploratory review.

Please, land this at your leisure 👍

youknowriad commented:
@nb @sirreal I see this is approved but stale a bit, can we refresh it and merge it or is there anything is blocking?

sirreal commented:
I don't think there's anything blocking. I'll try to circle around and finish this in the next few days. @nb let me know if you'd prefer to handle it.

sirreal commented:
Rebased

ockham commented on packages/wordcount/src/stripConnectors.js:
```diff
@@ -1,14 +1,11 @@
  * @return {string} The manipulated text.
  */
 export default function stripConnectors( settings, text ) {
-	if ( settings.connectorRegExp ) {
-		return text.replace( settings.connectorRegExp, ' ' );
-	}
-	return text;
```
@sirreal Is it possible that we botched the rebase? Seems like we lost that conditional (and I'm not sure we're comfortable relying _just_ on TS to make sure the `connectorRegExp` attr is there, since this will continue to be used from JS...)

Same for all the other `packages/wordcount/src/stripSomethings.js`

ockham replied:
I'll file a commit to fix.

ockham replied:
Ohh, nevermind, it wasn't us -- it was a deliberate decision to change these. Commit https://github.com/WordPress/gutenberg/pull/22077/commits/2dc5c11cddb60b9af93750460420b535e761e7be has more information.

ockham commented on packages/wordcount/src/index.js:
```diff
@@ -50,56 +61,56 @@ function loadSettings( type, userSettings ) {
+ * @param {WPWordCountSettings} settings Settings object containing regular expressions for each strip function
  *
- * @return {Array|{index: number, input: string}} The matched string.
+ * @return {number} The matched string.
```
```suggestion
 * @return {number} The number of words in the text.
```

(or `matches` instead of `words`).

ockham commented on packages/wordcount/src/index.js:
```diff
@@ -50,56 +61,56 @@ function loadSettings( type, userSettings ) {
- * @param {string} type         The type of count. Accepts ;words', 'characters_excluding_spaces', or 'characters_including_spaces'.
- * @param {Object} userSettings Custom settings object.
+ * @param {string}                  text         The text being processed
+ * @param {WPWordCountStrategy}     type	     The type of count. Accepts 'words', 'characters_excluding_spaces', or 'characters_including_spaces'.
```
```suggestion
 * @param {WPWordCountStrategy}     type         The type of count. Accepts 'words', 'characters_excluding_spaces', or 'characters_including_spaces'.
```

ockham committed:
Fix countWords return type label

ockham committed:
Fix some JSDoc whitespace

ockham reviewed (approved):
Approving. Merge pending CI :+1: 

I've read through your suggestions @sirreal but I'm afraid I'll need a bit longer to let them sink in :grimacing: and I'd rather get this un-stuck. We can iterate later if we decide that some types can use some more tightening :+1:
